src/OpenTelemetry/ProviderExtensions.cs

```diff
@@ -31,7 +32,7 @@ public static class ProviderExtensions
     /// </summary>
     /// <param name="baseProvider"><see cref="BaseProvider"/>.</param>
     /// <returns><see cref="Resource"/>if found otherwise <see cref="Resource.Empty"/>.</returns>
-    public static Resource GetResource(this BaseProvider baseProvider)
+    public static Resource GetResource([AllowNull] this BaseProvider baseProvider)
```

Comment:
Why are we allowing `null` values for `baseProvider`?

Reply:
It stems from this design:

```csharp
public abstract class BaseExporter<T> : IDisposable
{
   public BaseProvider? ParentProvider { get; internal set; }
}
```

We have `ParentProvider` on the base class but it is `null` until [added to the sdk provider](https://github.com/open-telemetry/opentelemetry-dotnet/blob/1f7f9315dc90e367ed38fcab411904e64536a802/src/OpenTelemetry/Trace/TracerProviderSdk.cs#L313).

In our exporters we have code that does things like [this.ParentProvider.GetResource()](https://github.com/open-telemetry/opentelemetry-dotnet/blob/1f7f9315dc90e367ed38fcab411904e64536a802/src/OpenTelemetry.Exporter.OpenTelemetryProtocol/OtlpLogExporter.cs#L85).

With the contract...

```csharp
public static Resource GetResource(this BaseProvider baseProvider)
```

...you'll get warning there about `ParentProvider` possibly being `null` when passed to that method.

`GetResource` actually handles `null` just fine: https://github.com/open-telemetry/opentelemetry-dotnet/blob/1f7f9315dc90e367ed38fcab411904e64536a802/src/OpenTelemetry/ProviderExtensions.cs#L53

Essentially it is a weird design that has put us in a corner 🤣 There's a bunch of things we _could_ do here. Convert the method to throw on `null` or unknown provider type (probably breaking). Suppress at the call sites using `!` (kind of ugly). Or change the signature to allow `null` somehow.

I went with the `[AllowNull]` style because it removes the warnings but doesn't promote the use of `null` to `GetResource` like `Resource GetResource(this BaseProvider? baseProvider)` style would. I thought `[AllowNull]` would better indicate: "You shouldn't pass something which is `null` to this method but you may if you really have a need to."